presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java

```diff
@@ -502,6 +505,39 @@ public PlanNode visitFilter(FilterNode node, RewriteContext<Set<VariableReferenc
             return new FilterNode(node.getSourceLocation(), node.getId(), node.getStatsEquivalentPlanNode(), source, node.getPredicate());
         }
 
+        @Override
+        public PlanNode visitCteConsumer(CteConsumerNode node, RewriteContext<Set<VariableReferenceExpression>> context)
+        {
+            // Some output can be pruned but current implementation of PhysicalCteProducer does not allow cteconsumer pruning
+            return node;
+        }
+
+        @Override
+        public PlanNode visitCteProducer(CteProducerNode node, RewriteContext<Set<VariableReferenceExpression>> context)
+        {
+            Set<VariableReferenceExpression> expectedInputs = new HashSet<>();
+            expectedInputs.addAll(node.getOutputVariables());
+            PlanNode source = context.rewrite(node.getSource(), expectedInputs);
+            return new CteProducerNode(node.getSourceLocation(), node.getId(), source, node.getCteName(), node.getRowCountVariable(), node.getOutputVariables());
+        }
+
+        @Override
+        public PlanNode visitSequence(SequenceNode node, RewriteContext<Set<VariableReferenceExpression>> context)
+        {
+            ImmutableSet.Builder<VariableReferenceExpression> cteProducersBuilder = ImmutableSet.builder();
+            node.getCteProducers().forEach(leftSource -> cteProducersBuilder.addAll(leftSource.getOutputVariables()));
+
+            Set<VariableReferenceExpression> leftInputs = cteProducersBuilder.build();
+
+            ImmutableSet.Builder<VariableReferenceExpression> primarySourceBuilder = ImmutableSet.builder();
+            primarySourceBuilder.addAll(node.getPrimarySource().getOutputVariables());
+            Set<VariableReferenceExpression> rightInputs = primarySourceBuilder.build();
+
+            List<PlanNode> cteProducers = node.getCteProducers().stream().map(leftSource -> context.rewrite(leftSource, leftInputs)).collect(Collectors.toList());
```

Comment:
Consider moving up to immediately below the collection of leftInputs, so that the same logic code block stay closer.

Reply:
context.rewrite(leftSource, ImmutableSet.of()) is enough? leftInputs seems to be redundant

Reply:
Need to populate the rewrite context with the current anticipated variables that the visitor aims to extract from the children.